### PR TITLE
Add bcomplex32 (complex<bfloat16>) ExtendedDType support

### PR DESCRIPTION
**Pull request by DeanTMaxim**
## Description

Adds `bcomplex32` (complex<bfloat16>) as a new ExtendedDType. Physical representation is a `bfloat16` array with a trailing dimension of size 2 holding `[real, imag]`, similar to how JAX represents `key<...>` dtypes.

**Motivation.** Complex bfloat16 enables ~2x memory reduction and bf16 tensor-core acceleration for workloads that need complex arithmetic but tolerate reduced precision (e.g. quantum circuit simulation, ML training).

One might ask why ExtendedDType is needed when `ml_dtypes` already provides `bcomplex32` as a NumPy dtype — the reason is XLA. XLA does not have `complex<bf16>` as an element type, so normal lowering produces `tensor<Nxcomplex<bf16>>` which StableHLO rejects. ExtendedDType rewrites the logical `tensor<Nxbcomplex32>` to physical `tensor<Nx2xbf16>` before it reaches XLA (same approach as PRNG keys, which are uint32 pairs under the hood).

## Status

🚧 Draft. `bcomplex32_edtype` is intentionally kept out of the global type promotion lattice (similar to PRNG keys), so cross-dtype operations require explicit decomposition via `lax.real` / `lax.imag` rather than silent promotion. This is consistent with how other ExtendedDTypes (prng, Pallas fuse dtype) behave in JAX.

CI will fail until `ml_dtypes >= 0.5.5` is released (see Dependencies).

## Implementation

- `jax._src.dtypes._BComplex32DType(ExtendedDType)` — dtype registration, `bcomplex32_edtype` singleton, `to_complex_dtype`.
- `jax._src.lax.lax_bcomplex32.BComplex32Rules` — ExtendedDType rules and primitive lowerings for arithmetic, matmul, transcendentals.
- `jax/_src/lax/lax_bcomplex32_array.BComplex32Array` — Python wrapper modeled after `PRNGKeyArray`.
- Updates to `lax.py`, `ufuncs.py`, `mlir.py`, `api.py` for integration.

**Matmul decomposition.** Complex matmul `C = A @ B` lowers to four real bf16 `dot_general` calls so it runs on bf16 tensor cores:

\`\`\`
C_re = A_re@B_re - A_im@B_im
C_im = A_re@B_im + A_im@B_re
\`\`\`

**Transcendentals.** `exp`, `log`, `sin`, `cos`, `tan`, `sqrt`, etc.
promote to complex64, compute, then demote back. Lossy but correct.

**Autodiff.** Supports `jit`, `grad`, `jvp`, `vjp`.

## Dependencies

This PR depends on `ml_dtypes >= 0.5.5` (unreleased), which adds `bcomplex32` as a NumPy extended dtype. Tracked in [jax-ml/ml_dtypes#355](https://github.com/jax-ml/ml_dtypes/issues/355).

Until `ml_dtypes` ships a release containing bcomplex32, CI will fail to import. Marking as draft until then.

cc @seberg for visibility on the ml_dtypes side.

## Testing

50 new tests in `tests/lax_bcomplex32_edtype_test.py` covering dtype registration, type promotion behavior, casting, array creation, arithmetic, matmul, transcendentals, autodiff, and JIT. All 50 pass against `ml_dtypes @ main`:

\`\`\`
============================= 50 passed in 14.50s =============================
\`\`\`

## Checklist

- [x] Single self-contained commit
- [x] `ruff check` / `ruff format --check` pass on new files
- [x] Tests added and passing
- [x] CLA signed
- [ ] `ml_dtypes >= 0.5.5` released (blocking CI)

### CODE REVIEW

File: jax/_src/lax/lax_bcomplex32_array.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+      raise TypeError("len() of unsized object")
+    return self.shape[0]
+
+  def __repr__(self):
+    return f"BComplex32Array(shape={self.shape}, dtype={self.dtype})"
+
+  def __iter__(self):
+    if self.ndim == 0:
+      raise TypeError("iteration over a 0-d array")
+    raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (DeanTMaxim) [reply jax/_src/lax/lax_bcomplex32_array.py]: __getitem__ is already wired up via `_set_array_base_attributes` (it forwards to the underlying `_base_array`'s `__getitem__`, which naturally preserves the trailing physical dim of size 2). Verified in test_indexing_and_iteration — both `arr[i]` and `arr[1:]` work.

Fixed __iter__ in ee07aed51.

**Comment by jakevdp** [comment]
Hi - an overall comment: since `ml_dtypes` already defines the `bcomplex32` dtype directly, I think we could add this sort of support without the extended dtype infrastructure.

**Comment by DeanTMaxim** [comment]
> Hi - an overall comment: since `ml_dtypes` already defines the `bcomplex32` dtype directly, I think we could add this sort of support without the extended dtype infrastructure.

The issue is that XLA doesn't support complex<bf16> as an element type. If bcomplex32 goes through the normal lowering path, the generated StableHLO has `tensor<Nxcomplex<bf16>>`, and XLA rejects it:

'stablehlo.convert' op result #0 must be complex type with 32/64-bit float elements, but got 'tensor<Nxcomplex<bf16>>'

So we need something that rewrites the logical shape before it hits XLA. ExtendedDType does exactly this (same pattern as PRNG keys — they're also a non-trivial physical repr under the hood). The alternative would be patching every primitive lowering individually, which seems worse.

If there's a cleaner approach I'm missing though, happy to reconsider.

**Comment by DeanTMaxim** [comment]
Quick update: `ml_dtypes` 0.6.0 now includes `bcomplex32`, so the dependency blocker is resolved.

The motivation for ExtendedDType still applies because StableHLO currently rejects `complex<bf16>`. However, this branch is far behind main and needs substantial rework.

Before rebuilding it, would ExtendedDType be an acceptable representation for numeric bcomplex32 in JAX? Also, should bcomplex32 be explicitly requested so that existing behavior such as `lax.complex(bf16, bf16) -> complex64` remains unchanged?

If so, I can restart from current main and split the dtype plumbing from the operation support.
